libs/s25main/figures/nofMiner.cpp

```diff
@@ -71,20 +72,100 @@ helpers::OptionalEnum<GoodType> nofMiner::ProduceWare()
 }
 
-bool nofMiner::StartWorking()
+unsigned int nofMiner::GetAddonSetting() const 
 {
-    MapPoint resPt = FindPointWithResource(GetRequiredResType());
-    if(!resPt.isValid())
-        return false;
     const GlobalGameSettings& settings = world->GetGGS();
-    bool inexhaustibleRes = settings.isEnabled(AddonId::INEXHAUSTIBLE_MINES)
-                            || (workplace->GetBuildingType() == BuildingType::GraniteMine
-                                && settings.isEnabled(AddonId::INEXHAUSTIBLE_GRANITEMINES));
-    if(!inexhaustibleRes)
-        world->ReduceResource(resPt);
+
+    switch(workplace->GetBuildingType())
+    {
+        case BuildingType::GoldMine: return settings.getSelection(AddonId::MINES_GOLD);
+        case BuildingType::IronMine: return settings.getSelection(AddonId::MINES_IRON);
+        case BuildingType::CoalMine: return settings.getSelection(AddonId::MINES_COAL);
+        case BuildingType::GraniteMine: return settings.getSelection(AddonId::MINES_GRANITE);
+        default: return 0;
+    }
+}
+
+bool nofMiner::StartWorking()
+{
+    workplace->is_emptyCycle = false;
+
+    // needs to have at least one resource spot
+    // 0 = is exhaustible
+    // 1 = settlers IV like
+    // 2 = inexhaustible
+    // 3 = everywhere
+    unsigned int addonSettings = GetAddonSetting();
+
+    switch (addonSettings)
+    {
+        case 1: // settlers IV style
+        {
+            int sumResAmount = 0;
+            MapPoint useResPt;
+
+            std::vector<MapPoint> resPts = FindAllPointsWithResource(GetRequiredResType());
+
+            // iterate over all points
+            for each(MapPoint curPt in resPts)
+            {
+                // calculate the absolute amount of resource beneath
+                uint8_t resAmount = world->GetNode(curPt).resources.getAmount();
+                sumResAmount += resAmount;
+
+                // if there is one with more than 1 quantity, keep it (for reducing)
+                if(resAmount > 1 && !useResPt.isValid())
+                    useResPt = curPt;
+            }
+
+            // no resources left (mine was built on invalid spot)
+            if(sumResAmount == 0)
+                return false;
+
+            // 19 = amount of nodes a mine can reach
+            // 7  = maximum resource amount a node possibly has
+            if(RANDOM.Rand(RANDOM_CONTEXT(), 19 * 7) > sumResAmount)
+            {
+                // failed, use food and start working - but produce nothing
+                workplace->is_emptyCycle = true;
+            } else
+            {
+                // if success, use 1 quantity if any
+                if(!useResPt.isValid())
+                    world->ReduceResource(useResPt);
+            }
+        }
+        break;
+        case 2: // inexhaustible
+        {
+            MapPoint resPt = FindPointWithResource(GetRequiredResType());
+            if(!resPt.isValid())
```

Comment:
Do we even need to check this here? This is already done in `AreWaresAvailable` and for inexhaustible it cannot disappear, can it?
If this is required I'd prefer: 
```suggestion
            if(!FindPointWithResource(GetRequiredResType()).isValid())
```
Again: No need for a named entity (variable) if it isn't (really) used.

Reply:
As far as I currently understand the logic, the ```AreWaresAvailable``` function is called only for any work cycle **except** the initial one. Maybe I missunderstood the logic at one point (or rather my testing, as first my mines ran exactly 1 time and then failed in the ```FindPointWithResource``` function in ```AreWaresAvailable```)
So yeah, going to use your suggestion :)

Reply:
The `AreWaresAvailable` is called for every work cycle, as that is the precondition for starting to work. See `TryToWork`.

The current work cycle is:
- If wares available: Wait_1 else Stop
- Work
- Wait_2
- Carry out ware (if any)
- Once back in the building (or if never left as no ware was produced): Restart